### PR TITLE
Add cloudwatch metric name whitelisting feature [Fixes #71]

### LINKED ISSUE #71: Filter metrics sent to cloudwatch
When using `CLOUDWATCH_ON=true`, it would be helpful to have some mechanism for filtering the metrics with a whitelist or blacklist. For example, we use the lag metrics heavily but don't have a lot of use for logend or offset metrics. Also, in some cases we want individual partition metrics and other times not. It may become a bit cost-prohibitive to continue sending all metrics to cloudwatch.

### PR DESCRIPTION
**Pull request by dlwodnjs42**
- Added functionality to cloudwatch integration: allows for whitelisting metric names rather than sending all 114 metrics.
- Allows for logs to catch logging from the RemorApp file
- Added a few functionality tests for overriding matches in metricfilter class.

### CODE REVIEW

**Comment by ohaibbq** [comment]
👍

File: README.md [hunk 1]
```diff
@@ -473,6 +473,8 @@ The following environment variables can be used to configure reporting to Cloudw
 
 * **CLOUDWATCH_ON** - `default false` reports metricsRegistry to cloudwatch, TO_REGISTRY will need to be switched on!
 * **CLOUDWATCH_NAME** - `default 'remora'` name to appear on cloudwatch
+* **WHITELIST** - `default ''` metric names to whitelist. The whitelist will be compared with metric names and will allow based on "equality" or "endswith". I.E "gauge.DATA_SAMPLE.lag" will be whitelisted if the environment variable is "DATA_SAMPLE.lag
```
Comment (soceanainn) [inline README.md:476]: In general, we're moving away from using whitelist as a term in our code. It would be better to call this what it actually is - i.e. `METRICS_FILTER` or `CLOUDWATCH_METRICS_FILTER`

Comment (soceanainn) [reply README.md:476]: Instead of a space delimited list of expression for 'endsWith' I think allowing users to provide a Regex expression here would allow for more configuration options

File: src/main/scala/RemoraApp.scala [hunk 2]
```diff
@@ -54,10 +55,28 @@ object RemoraApp extends App with nl.grons.metrics.scala.DefaultInstrumented wit
...
+        logger.debug("Metric Name:" + string)
+        if (filterString.length == 0) return true
+        for (v <- filterString) {
+            if (string.endsWith(v) || string == v ) {
+                return true
+            }
+        }
+        return false
+      }
+    }
```
Comment (soceanainn) [inline src/main/scala/RemoraApp.scala:73]: This code is duplicated in tests below, meaning that anyone can change this code here, breaking the desired behaviour, but tests would continue to pass. It would be better to pull this code out into it's own file, and implement it in a way that allows it to be used in your unit tests below

File: src/main/resources/application.conf [hunk 3]
```diff
@@ -2,7 +2,6 @@ akka {
   loggers = ["akka.event.slf4j.Slf4jLogger"]
   logging-filter = "akka.event.slf4j.Slf4jLoggingFilter"
   loglevel = "DEBUG"
-  stdout-loglevel = "OFF"
```
Comment (soceanainn) [inline src/main/resources/application.conf:5]: Why was this removed?

File: README.md (hunk 1 above)
Comment (dlwodnjs42) [reply README.md]: Changed to checking regex match rather than string equality or ends with and renamed it cloudwatch_metrics_filter

File: src/main/scala/RemoraApp.scala (hunk 2 above)
Comment (dlwodnjs42) [reply src/main/scala/RemoraApp.scala]: moved the code to a utils and used that function to test

File: src/main/resources/application.conf (hunk 3 above)
Comment (dlwodnjs42) [reply src/main/resources/application.conf:5]: re-added, was trying to figure out how to get logs from remora app since logger.info was actually not actually logging but was something else.

File: src/main/resources/logback.xml [hunk 4]
```diff
@@ -10,8 +10,16 @@
...
+        <appender-ref ref="stdout"/>
+    </root>
+
     <root level="${logback-root-level:-debug}">
-        <appender-ref ref="stdout" />
+        <appender name="stdout" class="ch.qos.logback.core.ConsoleAppender">
+            <encoder>
+                <pattern>%date{ISO8601} - [%level] - [%thread] %logger{36} - %message%n%ex</pattern>
+            </encoder>
+        </appender>
```
Comment (soceanainn) [inline src/main/resources/logback.xml:22]: Can you revert these changes?

File: src/main/scala/RemoraApp.scala [hunk 5]
```diff
@@ -7,10 +7,12 @@ import akka.stream.{ActorMaterializer, ActorMaterializerSettings, Supervision}
 import com.amazonaws.services.cloudwatch.{AmazonCloudWatchAsync, AmazonCloudWatchAsyncClientBuilder}
 import com.blacklocus.metrics.CloudWatchReporterBuilder
 import com.codahale.metrics.jvm.{GarbageCollectorMetricSet, MemoryUsageGaugeSet, ThreadStatesGaugeSet}
+import com.codahale.metrics.{MetricFilter, Metric}
 import com.typesafe.scalalogging.LazyLogging
 import config.{KafkaSettings, MetricsSettings}
 import kafka.admin.RemoraKafkaConsumerGroupService
 import reporter.RemoraDatadogReporter
+import Utils.buildMetricFilter
```
Comment (soceanainn) [inline src/main/scala/RemoraApp.scala:15]: Should be a lower case 'u' in utils

File: src/main/scala/Utils.scala [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import com.typesafe.scalalogging.LazyLogging
+import com.codahale.metrics.{Metric, MetricFilter}
+import scala.util.matching.Regex
+
+object Utils extends LazyLogging {
+    def buildMetricFilter(filterRegexPattern: String): MetricFilter = {
+        //since matches has is a string, the whitelist should be log filenames with commas with spaces
+        return new MetricFilter() {
+            override def matches(string: String, metric: Metric): Boolean =  {
+                logger.debug("Metric Name:" + string)
```
Comment (soceanainn) [inline src/main/scala/Utils.scala:11]: This debug log doesn't really help with anything, since it just adds a message with a random metric name

File: src/main/scala/Utils.scala [hunk 7]
```diff
@@ -0,0 +1,19 @@
+
+import com.typesafe.scalalogging.LazyLogging
+import com.codahale.metrics.{Metric, MetricFilter}
+import scala.util.matching.Regex
+
+object Utils extends LazyLogging {
+    def buildMetricFilter(filterRegexPattern: String): MetricFilter = {
+        //since matches has is a string, the whitelist should be log filenames with commas with spaces
```
Comment (soceanainn) [inline src/main/scala/Utils.scala:8]: Can you remove this comment, it's no longer relevant

File: src/main/scala/Utils.scala [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+        return new MetricFilter() {
+            override def matches(string: String, metric: Metric): Boolean =  {
+                logger.debug("Metric Name:" + string)
+                if (filterRegexPattern.length == 0 ||
+                    (new Regex(filterRegexPattern).findFirstIn(string).nonEmpty)) return true
```
Comment (soceanainn) [inline src/main/scala/Utils.scala:13]: I think it would be better to declare the Regex outside of this method so that it can be reused in subsequent calls, instead of creating a new Regex filter every time it's checking a metric name

File: src/main/scala/Utils.scala [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+                return false
```
Comment (soceanainn) [inline src/main/scala/Utils.scala:14]: The three 'return' statements used here aren't necessary

File: src/main/scala/Utils.scala [hunk 10]
```diff
@@ -0,0 +1,19 @@
+
+import com.typesafe.scalalogging.LazyLogging
+import com.codahale.metrics.{Metric, MetricFilter}
+import scala.util.matching.Regex
+
+object Utils extends LazyLogging {
```
Comment (soceanainn) [inline src/main/scala/Utils.scala:6]: Instead of calling this `utils` perhaps it could be called 'CloudWatchMetricFilter`

File: src/main/scala/RemoraApp.scala [hunk 11]
```diff
@@ -7,10 +7,12 @@ import akka.stream.{ActorMaterializer, ActorMaterializerSettings, Supervision}
 import com.amazonaws.services.cloudwatch.{AmazonCloudWatchAsync, AmazonCloudWatchAsyncClientBuilder}
 import com.blacklocus.metrics.CloudWatchReporterBuilder
 import com.codahale.metrics.jvm.{GarbageCollectorMetricSet, MemoryUsageGaugeSet, ThreadStatesGaugeSet}
+import com.codahale.metrics.{MetricFilter, Metric}
```
Comment (soceanainn) [inline src/main/scala/RemoraApp.scala:10]: I don't think these imports are needed anymore

File: src/main/resources/application.conf [hunk 12]
```diff
@@ -2,8 +2,8 @@ akka {
...
   log-dead-letters = off
+  stdout-loglevel = "OFF"
```
Comment (soceanainn) [inline src/main/resources/application.conf:6]: Can you move this back to line #5 so that it doesn't appear as a change

File: src/main/scala/config/MetricsSettings.scala [hunk 13]
```diff
@@ -22,7 +22,8 @@ object MetricsSettings {
       CloudWatch(
         config.getBoolean("metrics.cloudwatch.enabled"),
         config.getString("metrics.cloudwatch.name"),
-        config.getInt("metrics.cloudwatch.interval_minutes")
+        config.getInt("metrics.cloudwatch.interval_minutes"),
+        config.getString("metrics.cloudwatch.metric_filter")
```
Comment (soceanainn) [inline src/main/scala/config/MetricsSettings.scala:26]: Perhaps this would be nicer as an optional string, so that you can clean up the `length == 0` comparisons later

File: src/main/scala/RemoraApp.scala [hunk 14]
```diff
@@ -54,10 +56,14 @@ object RemoraApp extends App with nl.grons.metrics.scala.DefaultInstrumented wit
     logger.info("Reporting metricsRegistry to Cloudwatch")
     val amazonCloudWatchAsync: AmazonCloudWatchAsync = AmazonCloudWatchAsyncClientBuilder.defaultClient
 
+    // if null, cloudwatch grabs all the logs possible
```
Comment (soceanainn) [inline src/main/scala/RemoraApp.scala:59]: Can you remove the comment, I think the metricFilter code is easy enough to understand that it doesn't need an additional comment to explain it here

File: src/test/scala/reporter/CloudWatchReporterSpec.scala [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+
+  "Metrics filter" should "match any metric when empty string filter is given" in {
+    val filter = buildMetricFilter("")
+
+    filter.matches("any_metrics_name", metric) should be(true)
+    filter.matches("xfaewojz", metric) should be(true)
+  }
+
+  "Metrics filter" should "match metrics that have the regex pattern: true" in {
```
Comment (soceanainn) [inline src/test/scala/reporter/CloudWatchReporterSpec.scala:24]: Nitpick: you can just use `it` instead of `"Metrics filter"` for tests after the first one

File: src/test/scala/reporter/CloudWatchReporterSpec.scala [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    val filter = buildMetricFilter("([a-zA-Z]+.[a-zA-Z].*LFS+-loader-.+.lag)")
+
+    filter.matches("gauge.readings_V1-LFS-loader-aws.lag", metric) should be(true)
+  }
+
+   "Metrics filter" should "match metrics that have the regex pattern: false" in {
+    val filter = buildMetricFilter("([a-zA-Z]+.[a-zA-Z].*LFS+-loader-.+.lag)")
+
+    filter.matches("gauge.vers.stop.lag", metric) should be(false)
+  }
```
Comment (soceanainn) [inline src/test/scala/reporter/CloudWatchReporterSpec.scala:34]: Could you add a test for when an invalid regexp is passed as an input string?

File: src/test/scala/reporter/CloudWatchReporterSpec.scala [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+import scala.util.matching.Regex
+
+
+class CloudWatchReporterSpec extends FlatSpec with Matchers with PrivateMethodTester with MockFactory {
+
+  private val metricRegistry: MetricRegistry = new MetricRegistry
+  private val metric: Metric = mock[Metric]
+
+
+  "Metrics filter" should "match any metric when empty string filter is given" in {
```
Comment (soceanainn) [inline src/test/scala/reporter/CloudWatchReporterSpec.scala:17]: This should be `buildMetricFilter` instead of `Metrics filter` to make it easier to cross reference

File: src/test/scala/reporter/CloudWatchReporterSpec.scala [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+import com.blacklocus.metrics.CloudWatchReporterBuilder
+import com.codahale.metrics.{Metric, MetricFilter, MetricRegistry}
+import org.scalamock.scalatest.MockFactory
+import org.scalatest.{FlatSpec, Matchers, PrivateMethodTester}
+import Utils.buildMetricFilter
+import scala.util.matching.Regex
+
+
+class CloudWatchReporterSpec extends FlatSpec with Matchers with PrivateMethodTester with MockFactory {
```
Comment (soceanainn) [inline src/test/scala/reporter/CloudWatchReporterSpec.scala:11]: This class doesn't match the object that is containing the code (which is currently `Utils`)

File: src/main/scala/config/MetricsSettings.scala (hunk 13 above)
Comment (dlwodnjs42) [reply src/main/scala/config/MetricsSettings.scala:26]: i dont think config has an getoptionalstring method.

File: src/test/scala/reporter/CloudWatchReporterSpec.scala (hunk 16 above)
Comment (dlwodnjs42) [reply src/test/scala/reporter/CloudWatchReporterSpec.scala]: created a test of what happens if theres an invalid regexp since it would fail there instead of when its inserted to the method

File: src/main/scala/CloudWatchMetricFilter.scala [hunk 19]
```diff
@@ -0,0 +1,15 @@
+
+import com.typesafe.scalalogging.LazyLogging
+import com.codahale.metrics.{Metric, MetricFilter}
+import scala.util.matching.Regex
+
+object CloudWatchMetricFilter extends LazyLogging {
+    def buildMetricFilter(filterRegexPattern: Regex): MetricFilter = {
+        return new MetricFilter() {
+            override def matches(string: String, metric: Metric): Boolean =  {
+                return filterRegexPattern.findFirstIn(string).nonEmpty
```
Comment (soceanainn) [inline src/main/scala/CloudWatchMetricFilter.scala:10]: `return` isn't needed here or above

Comment (dlwodnjs42) [reply src/main/scala/CloudWatchMetricFilter.scala:10]: sorry its actually the first time coding in scala (have done java)
so should it be like

```
def buildMetricFilter(filterRegexPattern: Regex): MetricFilter = {
        new MetricFilter() {
            override def matches(string: String, metric: Metric): Boolean =  filterRegexPattern.findFirstIn(string).nonEmpty
       }
```

doesnt matches need a return value of a boolean and buildmetricfilter to have a return of a MetricFilter?

File: src/main/scala/RemoraApp.scala [hunk 20]
```diff
@@ -54,10 +56,13 @@ object RemoraApp extends App with nl.grons.metrics.scala.DefaultInstrumented wit
     logger.info("Reporting metricsRegistry to Cloudwatch")
     val amazonCloudWatchAsync: AmazonCloudWatchAsync = AmazonCloudWatchAsyncClientBuilder.defaultClient
 
+    val logMetricFilter = buildMetricFilter(new Regex(metricsSettings.cloudWatch.metricFilter))
```
Comment (soceanainn) [inline src/main/scala/RemoraApp.scala:59]: I think this can be simplified to `buildMetricFilter(metricsSettings.cloudWatch.metricFilter.r)`

File: src/main/scala/config/MetricsSettings.scala (hunk 13 above)
Comment (soceanainn) [inline src/main/scala/config/MetricsSettings.scala:26]: I think we can change the case class to be `metricFilter: Regex` and this line can become :
```
config.getString("metrics.cloudwatch.metric_filter").r
```
The r method is provided implicitly by the Regex object so you can do `"xyz".r` instead of `new Regex("xyz")` like you have below - this way we're validating the value when parsing the config instead of when building the filter.

File: src/test/scala/CloudWatchMetricFilterSpec.scala [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class CloudWatchMetricFilterSpec extends FlatSpec with Matchers with PrivateMethodTester with MockFactory {
+
+  private val metricRegistry: MetricRegistry = new MetricRegistry
+  private val metric: Metric = mock[Metric]
+
+
+  "buildMetricsFilter" should "match any metric when empty string filter is given" in {
+    val filter = buildMetricFilter(new Regex(""))
```
Comment (soceanainn) [inline src/test/scala/CloudWatchMetricFilterSpec.scala:18]: Similar to above, you can use `"".r` instead of `new Regex("")` just looks a little bit nicer 😃

File: src/main/scala/CloudWatchMetricFilter.scala (hunk 19 above)
Comment (soceanainn) [reply src/main/scala/CloudWatchMetricFilter.scala:10]: Scala automatically returns the value of the last statement for you so you don't have to use the 'return' keyword

Comment (dlwodnjs42) [reply src/main/scala/CloudWatchMetricFilter.scala:10]: wait thats pretty cool! Thanks for the tip!

Comment (soceanainn) [reply src/main/scala/CloudWatchMetricFilter.scala:10]: I think this can be simplified to:
```
class CloudWatchMetricFilter(filterRegexPattern: Regex) extends MetricFilter {
  override def matches(string: String, metric: Metric): Boolean =  {
    filterRegexPattern.findFirstIn(string).nonEmpty
  }
}```

Comment (dlwodnjs42) [reply src/main/scala/CloudWatchMetricFilter.scala]: scala is interesting 👍 :')

Comment (soceanainn) [reply src/main/scala/CloudWatchMetricFilter.scala]: Can you update the object to be a class instead as shown above?

**Comment by soceanainn** [comment]
👍

File: src/test/scala/CloudWatchMetricFilterSpec.scala [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    filter.matches("gauge.readings_V1-LFS-loader-aws.lag", metric) should be(true)
+  }
+
+   "it" should "match metrics that have the regex pattern: false" in {
+    val filter = new CloudWatchMetricFilter("([a-zA-Z]+.[a-zA-Z].*LFS+-loader-.+.lag)".r)
+
+    filter.matches("gauge.vers.stop.lag", metric) should be(false)
+  }
+
+   "it" should "error when theres an invalid regex pattern created" in {
```
Comment (soceanainn) [inline src/test/scala/CloudWatchMetricFilterSpec.scala:36]: This test case is no longer needed because `CloudWatchMetricFilter` is no longer doing the conversion from String to Regex

Comment (dlwodnjs42) [reply src/test/scala/CloudWatchMetricFilterSpec.scala]: 👍

**Comment by regispl** [comment]
👍

**Comment by soceanainn** [comment]
👍
